### PR TITLE
rpc: change rpc.Context.conns to a sync.Map

### PR DESCRIPTION
**Pull request by petermattis**
sync.Map is a concurrent map with amortized-constant-time loads, stores,
and deletes. It is safe for multiple goroutines to call a Map's methods
concurrently. sync.Map will be part of the go1.9 stdlib.

Use sync.Map for rpc.Context.conns.

### CODE REVIEW

File: pkg/util/syncutil/map.go [hunk 1]
```diff
@@ -0,0 +1,371 @@
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// TODO(peter): This file was copied from sync/map.go in the Go stdlib. The
```
Comment (tamird) [inline pkg/util/syncutil/map.go]: Haven't reviewed in detail, but I believe you can use x/sync instead of adding this copy.

**Comment by petermattis** [comment]
`rpc.Context.conns.Mutex` has shown up on mutex profiles, but I've been unable to demonstrate a performance benefit from this change.

There is no urgency to merging this, but it shows how `sync.Map` could be used to replace other use of mutex protected maps such as in `storage.RaftTransport` and possibly `storage.Store.mu.replicas`.

File: pkg/util/syncutil/map.go (hunk 1 above)
Comment (petermattis) [inline pkg/util/syncutil/map.go]: Good point. Changed.

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/rpc/context.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-KntAUM55q756T_xbL7R:-KntAUM55q756T_xbL7S:bi8b8rr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fb19b21135808231754b0db9c758fc2d801d4/pkg/rpc/context.go#L136)):*
> ```Go
> }
> 
> type errValue struct {
> ```

is this structure just used to avoid storing `nil` in the atomic `heartbeatErr`? might be nicer to use unsafe.Pointer, but otherwise this needs a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-:-KntA87aAT1S_MVkJAqA:bnfp4nl)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/rpc/context.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-KntAUM55q756T_xbL7R:-KntUPXNXOV6_xB2EJpT:b-1wdtvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fb19b21135808231754b0db9c758fc2d801d4/pkg/rpc/context.go#L136)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this structure just used to avoid storing `nil` in the atomic `heartbeatErr`? might be nicer to use unsafe.Pointer, but otherwise this needs a comment.
</blockquote></details>

Yes. An `atomic.Value` cannot store `nil`. How would we use `unsafe.Pointer` instead of `atomic.Value`? Note than `error` is an interface (i.e. a 2-word struct) and not a pointer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/rpc/context.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-KntAUM55q756T_xbL7R:-KntUsIhMtmGnWUO69r_:b-upb16z) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fb19b21135808231754b0db9c758fc2d801d4/pkg/rpc/context.go#L136)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes. An `atomic.Value` cannot store `nil`. How would we use `unsafe.Pointer` instead of `atomic.Value`? Note than `error` is an interface (i.e. a 2-word struct) and not a pointer.
</blockquote></details>

Right, I was imagining that it'd store a *error. It's awkward either way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807)*

**Comment by petermattis** [comment]
Despite the lack of measured perf benefit, do you think this should go in? Note that `DistSender` calls `GRPCDial` for each replica of a range, even if the RPC is only sent to one of the replicas. Perhaps I should write a small benchmark for `GRPCDial`.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/rpc/context.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-KntAUM55q756T_xbL7R:-KntVLPfhw_uhgnuDkq1:b1gcp69) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fb19b21135808231754b0db9c758fc2d801d4/pkg/rpc/context.go#L136)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Right, I was imagining that it'd store a *error. It's awkward either way.
</blockquote></details>

Ack, that would work. I'm going to leave as-is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-:-KntVRPWs0TmLsC_Ftyp:b-blohes)*

**Comment by tamird** [comment]
I think a benchmark would be worthwhile.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/rpc/context.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-KntAUM55q756T_xbL7R:-KntWQ05LS8J9IekrUYl:b90wu5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fb19b21135808231754b0db9c758fc2d801d4/pkg/rpc/context.go#L136)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ack, that would work. I'm going to leave as-is.
</blockquote></details>

Sure, just add a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-:-KntXAW6u2c9TgVbinEc:b-w8m7v)*

**Comment by petermattis** [comment]
Benchmark added. That's more of a speedup than I would have guessed (see commit message).

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/rpc/context.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-KntAUM55q756T_xbL7R:-Knta77-kF7TgO-0Kr8J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fb19b21135808231754b0db9c758fc2d801d4/pkg/rpc/context.go#L136)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Sure, just add a comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-:-Knta8sdWEjsHxSvak2s:b-jy5ypz)*

**Comment by tamird** [comment]
Neat.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16807#-:-KntaxHMX2aeVI8DnVd7:b-83v4qp)*

**Comment by petermattis** [comment]
> Neat.

Agreed. I'm going to take a look at changing some of the other read-mostly maps such as in `RaftTransport` and `Store.mu.replicas`.

**Review by a-robinson — APPROVED** [review]
Woah, that speedup is way bigger than I would have thought.

**Pull request merged**